tensorflow_addons/seq2seq/python/attention_wrapper.py

```diff
@@ -0,0 +1,1874 @@
+  # alignments shape is
+  #   [batch_size, 1, memory_time]
+  # attention_mechanism.values shape is
+  #   [batch_size, memory_time, memory_size]
+  # the batched matmul is over memory_time, so the output shape is
+  #   [batch_size, 1, memory_size].
+  # we then squeeze out the singleton dim.
+  context_ = math_ops.matmul(expanded_alignments, attention_mechanism.values)
+  context_ = array_ops.squeeze(context_, [1])
+
+  if attention_layer is not None:
+    attention = attention_layer(array_ops.concat([cell_output, context_], 1))
+  else:
+    attention = context_
+
+  return attention, alignments, next_attention_state
+
+
+class AttentionWrapper(rnn_cell_impl.RNNCell):
+  """Wraps another `RNNCell` with attention.
+  """
+
+  def __init__(self,
+               cell,
+               attention_mechanism,
+               attention_layer_size=None,
+               alignment_history=False,
+               cell_input_fn=None,
+               output_attention=True,
+               initial_cell_state=None,
+               name=None,
+               attention_layer=None,
+               attention_fn=None):
+    """Construct the `AttentionWrapper`.
+
+    **NOTE** If you are using the `BeamSearchDecoder` with a cell wrapped in
+    `AttentionWrapper`, then you must ensure that:
+
+    - The encoder output has been tiled to `beam_width` via
+      `tf.contrib.seq2seq.tile_batch` (NOT `tf.tile`).
+    - The `batch_size` argument passed to the `zero_state` method of this
+      wrapper is equal to `true_batch_size * beam_width`.
+    - The initial state created with `zero_state` above contains a
+      `cell_state` value containing properly tiled final state from the
+      encoder.
+
+    An example:
+
+    ```
+    tiled_encoder_outputs = tf.contrib.seq2seq.tile_batch(
+        encoder_outputs, multiplier=beam_width)
+    tiled_encoder_final_state = tf.conrib.seq2seq.tile_batch(
+        encoder_final_state, multiplier=beam_width)
+    tiled_sequence_length = tf.contrib.seq2seq.tile_batch(
+        sequence_length, multiplier=beam_width)
+    attention_mechanism = MyFavoriteAttentionMechanism(
+        num_units=attention_depth,
+        memory=tiled_inputs,
+        memory_sequence_length=tiled_sequence_length)
+    attention_cell = AttentionWrapper(cell, attention_mechanism, ...)
+    decoder_initial_state = attention_cell.zero_state(
+        dtype, batch_size=true_batch_size * beam_width)
+    decoder_initial_state = decoder_initial_state.clone(
+        cell_state=tiled_encoder_final_state)
+    ```
+
+    Args:
+      cell: An instance of `RNNCell`.
+      attention_mechanism: A list of `AttentionMechanism` instances or a single
+        instance.
+      attention_layer_size: A list of Python integers or a single Python
+        integer, the depth of the attention (output) layer(s). If None
+        (default), use the context as attention at each time step. Otherwise,
+        feed the context and cell output into the attention layer to gen
```

Comment:
Is there a plan in the future to utilize autograph and TF2 semantics? This is a ton of code so I understand it not happening right away, but I do hope that the code in addons sets a good example of TF2 code, seeing as we only support TF2.x

Reply:
Hmm I must have misclicked, but this should have pointed to the `ops.control_dependancies` call

Reply:
autograph should be something on the road map, i think it is still in the alpha stage. I can update the code once it is mature enough.